File: src/main/java/com/dsb/rest/dao/EquipementsDAO.java
package com.dsb.rest.dao;

import com.dsb.rest.model.Equipment;
import com.dsb.rest.model.Statistic;
import com.fasterxml.jackson.databind.ObjectMapper;
import org.springframework.lang.Nullable;
import org.springframework.stereotype.Repository;

import java.io.FileReader;
import java.io.IOException;
import java.util.Arrays;
import java.util.Collections;
import java.util.Comparator;
import java.util.List;
import java.util.regex.Pattern;
import java.util.stream.Collector;
import java.util.stream.Collectors;

@Repository
public class EquipementsDAO {
    private static final int pageSize = 50;

    private List<Equipment> equipmentListSort = null;

    private String name = null;

    private List<String> lastFilteredStats = null;

    private List<Equipment> deserialized(String dir) throws IOException {
        ObjectMapper objectMapper = new ObjectMapper();
        Equipment[] equipment = objectMapper.readValue(new FileReader(dir), Equipment[].class);
        return Arrays.asList(equipment);
    }

    public List<Equipment> getAllEquipements(String dir, int page) throws IOException {
        List<Equipment> equipmentList = this.deserialized(dir);

        equipmentList = equipmentList.stream().sorted(Comparator.comparingInt((Equipment e) -> Integer.parseInt(e.getLevel()))).collect(toListReversed());
        return equipmentList.stream().skip(pageSize * (page - 1)).limit(pageSize).collect(Collectors.toList());
    }

    public Integer getTotalEquipement(String dir) throws IOException {
        List<Equipment> equipmentList = this.deserialized(dir);
        return equipmentList.size();
    }

    public List<Equipment> filter(String dir, int page, int level, @Nullable String name, @Nullable List<String> stats) throws IOException {
        if (name != null) {
            return this.filterEquipementsByName(dir, page, level, name);
        } else if (stats != null && stats.size() >= 1) {
            return this.filterEquipementsByStat(dir, page, level, stats);
        } else {
            return this.filterEquipementsByLevel(dir, page, level);
        }
    }

    public List<Equipment> filterEquipementsByLevel(String dir, int page, int level) throws IOException {
        List<Equipment> equipmentList;

        if (equipmentListSort != null) {
            equipmentList = equipmentListSort;
        } else {
            equipmentList = this.deserialized(dir);
        }

        equipmentListSort = equipmentList
                .stream()
                .filter(equipments -> Integer.parseInt(equipments.getLevel()) >= level)
                .skip(pageSize * (page - 1))
                .limit(pageSize)
                .collect(Collectors.toList());

        return equipmentListSort;
    }

    public List<Equipment> filterEquipementsByName(String dir, int page, int level, String name) throws IOException {
        List<Equipment> equipmentList = this.deserialized(dir);

        equipmentListSort = equipmentList
                .stream()
                .filter(equipments -> Pattern.compile(Pattern.quote(name), Pattern.CASE_INSENSITIVE).matcher(equipments.getName()).find())
                .filter(equipments -> Integer.parseInt(equipments.getLevel()) <= level)
                .skip(pageSize * (page - 1))
                .limit(pageSize)
                .collect(toListReversed());

        return equipmentListSort;
    }

    public List<Equipment> filterEquipementsByStat(String dir, int page, int level, List<String> stats) throws IOException {
        List<Equipment> equipmentList;
        if (equipmentListSort != null) {
            if (stats.size() < this.lastFilteredStats.size()) {
                equipmentList = this.deserialized(dir);
            } else {
                equipmentList = equipmentListSort;
            }
        } else {
            equipmentList = this.deserialized(dir);
        }

        this.lastFilteredStats = stats;

        stats.forEach(s -> equipmentListSort = equipmentList
                .stream()
                .filter(equipments -> equipments.getStatistics().stream().anyMatch(Statistic.getPredicateByLabel(s)))
                .filter(equipments -> Integer.parseInt(equipments.getLevel()) <= level)
                .collect(Collectors.toList())
        );

        equipmentListSort = equipmentListSort
                .stream()
                .sorted(Comparator.comparingInt((Equipment e) -> Integer.parseInt(e.getLevel())))
                .collect(toListReversed());

        equipmentListSort = equipmentListSort
                .stream()
                .skip(pageSize * (page - 1))
                .limit(pageSize)
                .collect(Collectors.toList());

        return equipmentListSort;
    }

    private static <T> Collector<T, ?, List<T>> toListReversed() {
        return Collectors.collectingAndThen(Collectors.toList(), l -> {
            Collections.reverse(l);
            return l;
        });
    }
}

File: src/main/kotlin/com/dsb/rest/dao/EquipementsDAO.kt
package com.dsb.rest.dao

import com.dsb.rest.model.Equipement
import com.dsb.rest.model.Statistic
import com.fasterxml.jackson.databind.ObjectMapper
import com.fasterxml.jackson.module.kotlin.KotlinModule
import com.fasterxml.jackson.module.kotlin.readValue
import java.io.FileReader
import java.util.function.Consumer
import java.util.regex.Pattern
import java.util.stream.Collectors

class EquipementsDAO {

    companion object {
        private const val pageSize = 50

        private val mapper = ObjectMapper().registerModule(KotlinModule())

        private var equipmentListSort: List<Equipement>? = null
        private var lastFilteredStats: List<String>? = null
    }

    fun getAllEquipements(dir: String, page: Int): List<Equipement> {
        return deserialized(dir).sortedBy { it.level }
            .take(pageSize * (page * 1))
            .drop(pageSize * (page - 1))
            .asReversed()
    }

    fun getTotalEquipement(dir: String): Int {
        val equipmentList = deserialized(dir)
        return equipmentList.size
    }

    fun filter(
        dir: String,
        page: Int,
        level: Int,
        name: String?,
        stats: List<String>
    ): List<Equipement> {
        return when {
            name != null -> filterEquipementsByName(dir, page, level, name)
            stats.isNotEmpty() -> filterEquipementsByStat(dir, page, level, stats)
            else -> filterEquipementsByLevel(dir, page, level)
        }
    }

    fun filterEquipementsByLevel(dir: String, page: Int, level: Int): List<Equipement> {
        val equipmentList: List<Equipement> = if (equipmentListSort != null) {
            equipmentListSort!!
        } else {
            deserialized(dir)
        }
        equipmentListSort = equipmentList
            .filter { it.level.toInt() <= level }
            .take(pageSize * (page * 1))
            .drop(pageSize * (page - 1))
            .asReversed()
        return equipmentListSort!!
    }

    fun filterEquipementsByName(dir: String, page: Int, level: Int, name: String): List<Equipement> {
        equipmentListSort = deserialized(dir)
            .filter { Pattern.compile(Pattern.quote(name), Pattern.CASE_INSENSITIVE).matcher(it.name).find() }
            .filter { it.level.toInt() <= level }
            .take(pageSize * (page * 1))
            .drop(pageSize * (page - 1))
            .asReversed()
        return equipmentListSort!!
    }

    fun filterEquipementsByStat(dir: String, page: Int, level: Int, stats: List<String>): List<Equipement> {
        val equipmentList: List<Equipement> = if (equipmentListSort != null) {
            if (stats.size < lastFilteredStats!!.size) {
                deserialized(dir)
            } else {
                equipmentListSort!!
            }
        } else {
            deserialized(dir)
        }
        lastFilteredStats = stats
        stats.forEach(
            Consumer { s: String ->
                equipmentListSort = equipmentList
                    .stream()
                    .filter { (_, _, _, _, statistics) ->
                        statistics.stream().anyMatch(Statistic.getPredicateByLabel(s))
                    }
                    .filter { (_, _, level1) -> level1.toInt() <= level }
                    .collect(Collectors.toList())
            }
        )
        equipmentListSort = equipmentListSort!!
            .sortedBy { it.level }
            .take(pageSize * (page * 1))
            .drop(pageSize * (page - 1))
            .asReversed()
        return equipmentListSort!!.toList()
    }

    private fun deserialized(dir: String): List<Equipement> {
        return mapper.readValue(FileReader(dir))
    }
}